plugins/promela/hu.bme.mit.gamma.xsts.promela.transformation/src/hu/bme/mit/gamma/xsts/promela/transformation/serializer/ModelSerializer.xtend

```diff
@@ -40,39 +46,64 @@ class ModelSerializer {
 	protected final extension HavocHandler havocHandler = HavocHandler.INSTANCE
 	protected final extension ExpressionUtil expressionUtil = ExpressionUtil.INSTANCE
 	protected final extension ArrayHandler arrayHandler = ArrayHandler.INSTANCE
+	protected final extension GammaEcoreUtil ecoreUtil = GammaEcoreUtil.INSTANCE
 	
-	def String serializePromela(XSTS xSts) '''
-		«xSts.serializeDeclaration»
-		
-		byte flag = 0;
-		
+	protected Map<List<Action>, Integer> parallelMapping
+	protected Map<Action, List<VariableDeclaration>> parallelVariableMapping
+	protected int maxParallelNumber
+	
+	def String serializePromela(XSTS xSts) {
+		maxParallelNumber = 0
+		parallelMapping = newHashMap
+		parallelVariableMapping = newHashMap
 		
-		proctype EnvTrans() {
-			(flag > 0);
-		ENV:
-			atomic {
-				«xSts.environmentalAction.serialize»
-				flag = 2;
-			};
-			goto TRANS;
-		TRANS:
-			atomic {
-				«FOR transition : xSts.transitions»
-					«transition.action.serialize»
-				«ENDFOR»
-				flag = 1;
-			};
-			goto ENV;
+		val initializingActions = xSts.initializingAction
+		val environmentalActions = xSts.environmentalAction
+		val List<Action> actions = newArrayList
+		actions += initializingActions
+		actions += environmentalActions
+		val transitions = xSts.transitions
+		for (transition : transitions) {
+			actions += transition.action
 		}
 		
-		init {
-			atomic {
-				«xSts.initializingAction.serialize»
-				run EnvTrans();
-				flag = 1;
+		actions.createParallelMapping
+		
+		return '''
+			«xSts.serializeDeclaration»
+			
+			«serializeParallelChan»
+			byte flag = 0;
+			
+			«serializeParallelProcesses»
+			
+			proctype EnvTrans() {
+				(flag > 0);
+			ENV:
+				atomic {
+					«environmentalActions.serialize»
+					flag = 2;
+				};
+				goto TRANS;
+			TRANS:
+				atomic {
+					«FOR transition : transitions»
+						«transition.action.serialize»
```

Comment:
These transitions should be in an Or relation (only one of them should fire in a single cycle).